File: BinarySearch/Python/special_array_with_x_elements_greater_than_or_equal_x.py
class Solution:
    def specialArray(self, nums: list[int]) -> int:
        n = len(nums)
        nums.sort()
        cnt, idx = 0, 0
        while cnt < n + 1:
            if idx > n-1 or cnt > n - idx: return -1
            if cnt <= nums[idx] and cnt == n - idx:
                return cnt
            elif cnt > nums[idx]:
                idx += 1
            else:
                cnt += 1
        return -1

    # Binary Search approach
    def specialArrayBinarySearch(self, nums: list[int]) -> int:
        nums.sort()
        left, right = 0, len(nums)
        while left <= right:
            mid = (left + right) // 2
            cnt = sum(num >= mid for num in nums)
            if cnt == mid:
                return mid
            elif cnt < mid:
                right = mid - 1
            else:
                left = mid + 1
        return -1

File: BinarySearch/CPP/special_array_with_x_elements_greater_than_or_equal_x.cpp
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
    public:
        int specialArray(vector<int>& nums) {
            int n = nums.size(), cnt = 0, idx = 0;
            sort(nums.begin(), nums.end());
            while (cnt < n + 1) {
                if (idx > n - 1 || cnt > n - idx) return -1;
                if (cnt <= nums[idx] && cnt == n - idx) return cnt;
                else cnt > nums[idx] ? idx++ : cnt++;
            }
            return -1;
        }

        // Binary Search approach
        int specialArray(vector<int>& nums) {
            int right = nums.size(), left = 0;
            sort(nums.begin(), nums.end());
            while (left <= right) {
                int mid = left + (right - left) / 2;
                int cnt = nums.end() - lower_bound(nums.begin(), nums.end(), mid);
                if (mid == cnt) return mid;
                else if (mid <cnt) left = mid + 1;
                else right = mid - 1;
            }
            return -1;
        }
    };
